# Determination of loan borrowers' reliability

## Description

The customer is the credit department of the bank. It is necessary to understand whether the marital status and the number of children of the client affect the fact of repayment of the loan on time. Input data from the bank — statistics on the solvency of customers.

The results of the study will be taken into account when building a credit scoring model — a special system that evaluates the ability of a potential borrower to repay a loan to a bank.

**The purpose of the study** is to determine whether the primary parameters of borrowers affect the appearance of debt when paying off the loan and thus help to adjust the scoring model of the bank. In particular, it is necessary to test a number of hypotheses:

 - Is there a relationship between having children and repayment of the loan on time?
 - Is there a relationship between marital status and repayment of the loan on time?
 - Is there a relationship between the level of income and repayment of the loan on time?
 - How do different loan goals affect its repayment on time?

**Research plan**

Data on the reliability of borrowers is obtained from the data.csv file. Nothing is known about the quality of the data. Therefore, before testing hypotheses, we will need to review the data.
The data will be checked for errors and their impact on the study will be assessed. Then, at the preprocessing stage, the most critical data errors will be corrected.

Thus, the study will take place in three stages:
Data overview.
Data preprocessing.
Hypothesis testing.

___
## Data overview
___

Let's make a first idea of the data

Importing the pandas library - The main analytics tool

In [95]:
import pandas as pd

Read the data.csv file from the /datasets folder and save it in the df variable:

In [96]:
try:
    df = pd.read_csv('/Users/vladislav/Documents/datasets/data.csv')
except:
    df = pd.read_csv('/datasets/data.csv')

Let's display the first ten rows of the table:

In [97]:
df.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

We will also get general information about the table:

In [98]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


And we will conduct an upper-level numerical analysis of the data:

In [99]:
df.describe(include='all').T

count unique              top   freq           mean  \
children          21525.0    NaN              NaN    NaN       0.538908   
days_employed     19351.0    NaN              NaN    NaN   63046.497661   
dob_years         21525.0    NaN              NaN    NaN       43.29338   
education           21525     15          среднее  13750            NaN   
education_id      21525.0    NaN              NaN    NaN       0.817236   
family_status       21525      5  женат / замужем  12380            NaN   
family_status_id  21525.0    NaN              NaN    NaN       0.972544   
gender              21525      3                F  14236            NaN   
income_type         21525      8        сотрудник  11119            NaN   
debt              21525.0    NaN              NaN    NaN       0.080883   
total_income      19351.0    NaN              NaN    NaN  167422.302208   
purpose             21525     38          свадьба    797            NaN   

                            std           min            25%            50%  \
children               1.381587          -1.0            0.0            0.0   
days_employed     140827.311974 -18388.949901   -2747.423625   -1203.369529   
dob_years             12.574584           0.0           33.0           42.0   
education                   NaN           NaN            NaN            NaN   
education_id           0.548138           0.0            1.0            1.0   
family_status               NaN           NaN            NaN            NaN   
family_status_id       1.420324           0.0            0.0            0.0   
gender                      NaN           NaN            NaN            NaN   
income_type                 NaN           NaN            NaN            NaN   
debt                   0.272661           0.0            0.0            0.0   
total_income      102971.566448  20667.263793  103053.152913  145017.937533   
purpose                     NaN           NaN            NaN            NaN   

                            75%             max  
children                    1.0            20.0  
days_employed       -291.095954   401755.400475  
dob_years                  53.0            75.0  
education                   NaN             NaN  
education_id                1.0             4.0  
family_status               NaN             NaN  
family_status_id            1.0             4.0  
gender                      NaN             NaN  
income_type                 NaN             NaN  
debt                        0.0             1.0  
total_income      203435.067663  2265604.028723  
purpose                     NaN             NaN

So, in the table **11** columns. Data type in columns — **int64,float64,object**.

According to the data documentation:
* **children** — number of children in the family
* **days_employed** — total work experience in days
* **dob_years** — client's age in years
* **education** — the client's level of education
* **education_id** — education level identifier
* **family_status** — marital status
* **family_status_id** — id of marital status
* **gender** — gender
* **income_type** — type of employment
* **debt** — did I have a debt on repayment of loans
* **total_income** — monthly income
* **purpose** — purpose of obtaining a loan

There are no style violations in column names.

The number of values in the columns varies. So there are missing values in the data.

In the columns with numeric values **children,days_employed,dob_years** already at this stage, you can notice artifacts - a negative number of days of total experience and the number of children, "zero" age of borrowers.

**Conclusion**

Each row of the table contains data about the borrower. Most of the columns describe the social portrait of the borrower. The data in the debt column are key to the study - they answer the question about the credit quality of this borrower.
Previously, it can be argued that there is enough data to test hypotheses. But there are missing values in the data.
To move on, you need to process them.

___
## Data preprocessing
___
Let's fill missing values. Then we will check the data for duplicates.

### Missing values
Let's calculate how many missing values are in the table:

In [100]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Missing values do not affect the verification of all hypotheses. So in **total_income**, omissions are not important for testing hypotheses about the dependence of marital status and creditworthiness, but they are important for studying the dependence of creditworthiness on income level. Omissions in **days_employed** are not important for research.

In practice, it would be correct to establish the reason for the omissions and restore the data. There is no such possibility in the training project. 

We have to:

* fill in these gaps with explicit notation, 
* to assess how much they will damage the calculations.

It is impractical to delete lines with omissions in the **days_employed** column, since these lines contain useful information about the number of children, marital status and the purpose of the loan, which will allow answering other questions posed before the study. Replace the gaps in it with 0.

In [101]:
df['days_employed'] = df['days_employed'].fillna(0)

It is impossible to do the same with omissions in the **total_income** column, since in the future this may affect the consolidated income indicator for the entire column and, accordingly, conclusions about the dependence of creditworthiness on income.

It would be more correct to replace the gaps in **total_income** with the median value for the column.

In [102]:
median_income = df['total_income'].median()
df['total_income'] = df['total_income'].fillna(median_income)

Let's check that there are no more missing values:

In [103]:
df.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

The next step is to deal with the negative values in the **days_employed** column. Let's study for which type of employment negative values are displayed. Let's group the data by type of employment and find the maximum value of the length of service in days.

In [104]:
df.groupby('income_type')['days_employed'].max()

income_type
безработный        395302.838654
в декрете           -3296.759962
госслужащий             0.000000
компаньон               0.000000
пенсионер          401755.400475
предприниматель         0.000000
сотрудник               0.000000
студент              -578.751554
Name: days_employed, dtype: float64

It can be seen that positive values are characteristic only for "unemployed" population groups: "unemployed" and "retired". It seems that when uploading data for working groups of the population, a technical error occurred, displaying the length of service with a minus sign. If the task was to determine creditworthiness depending on the length of service, it would be necessary to clarify in the data source whether everything is really so. But since this change is not critical for research purposes, you can simply get the absolute value from the **days_employed** column to get rid of artifacts..

To do this, we will get the absolute value from the column **'days_employed'** and add it to the new column **'abs_days_employed'**:

In [105]:
df['abs_days_employed'] = df['days_employed'].abs()

Output the top 5 lines for verification:

In [106]:
#df['abs_days_employed'] = df['days_employed'].apply(absolute) #Initial version of the code
display(df.head())

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  abs_days_employed  
0               покупка жилья        8437.673028  
1     приобретение автомобиля        4024.803754  
2               покупка жилья        5623.422610  
3  дополнительное образование        4124.747207  
4             сыграть свадьбу      340266.072047

Absolute values are obtained. You can delete a column **days_employed**:

In [107]:
del df['days_employed']
display(df.head())

children  dob_years education  education_id     family_status  \
0         1         42    высшее             0   женат / замужем   
1         1         36   среднее             1   женат / замужем   
2         0         33   Среднее             1   женат / замужем   
3         3         32   среднее             1   женат / замужем   
4         0         53   среднее             1  гражданский брак   

   family_status_id gender income_type  debt   total_income  \
0                 0      F   сотрудник     0  253875.639453   
1                 0      F   сотрудник     0  112080.014102   
2                 0      M   сотрудник     0  145885.952297   
3                 0      M   сотрудник     0  267628.550329   
4                 1      F   пенсионер     0  158616.077870   

                      purpose  abs_days_employed  
0               покупка жилья        8437.673028  
1     приобретение автомобиля        4024.803754  
2               покупка жилья        5623.422610  
3  дополнительное образование        4124.747207  
4             сыграть свадьбу      340266.072047

### Duplicates

Let's check the number of possible duplicates in the column **children**:

In [108]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Due to the lack of an opportunity to clarify the information and compare it with the semantic values in other columns, we will replace the incorrect values:

In [109]:
try:
    df['children'] = df['children'].replace(-1,1)
    df['children'] = df['children'].replace(20,2)
except:
    print('Искомое значение не найдено')
    
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

Replacements were made, there were no errors during replacement.

Next, check for errors and duplicates in the column with the age value **dob_years**:

In [110]:
df['dob_years'].unique()

array([42, 36, 33, 32, 53, 27, 43, 50, 35, 41, 40, 65, 54, 56, 26, 48, 24,
       21, 57, 67, 28, 63, 62, 47, 34, 68, 25, 31, 30, 20, 49, 37, 45, 61,
       64, 44, 52, 46, 23, 38, 39, 51,  0, 59, 29, 60, 55, 58, 71, 22, 73,
       66, 69, 19, 72, 70, 74, 75])

There are incorrect values in the age data: the age of some people is not specified, the value is "0". Replace it with the median value.

In [111]:
df['dob_years'] = df['dob_years'].replace(0,df['dob_years'].median())
df['dob_years'].unique()

array([42., 36., 33., 32., 53., 27., 43., 50., 35., 41., 40., 65., 54.,
       56., 26., 48., 24., 21., 57., 67., 28., 63., 62., 47., 34., 68.,
       25., 31., 30., 20., 49., 37., 45., 61., 64., 44., 52., 46., 23.,
       38., 39., 51., 59., 29., 60., 55., 58., 71., 22., 73., 66., 69.,
       19., 72., 70., 74., 75.])

The replacement was made correctly. Now let's check the data on the gender of borrowers in the column **gender**:

In [112]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

As you can see, there is an artifact - "XNA". Since there are 2 times more women in the sample, the probability that the unspecified gender of the borrower was also female is higher. Let's add this value.

In [113]:
df['gender'] = df['gender'].replace('XNA','F')
df['gender'].value_counts()

F    14237
M     7288
Name: gender, dtype: int64

The replacement was made correctly.

Next, check the number of possible duplicates in the column **family_status**:

In [114]:
df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

There are no duplicates in this column. Also in the table we see the column **family_status_id**. Let's compare its values with the column**family_status**

In [115]:
df.groupby('family_status')['family_status_id'].mean()

family_status
Не женат / не замужем    4
в разводе                3
вдовец / вдова           2
гражданский брак         1
женат / замужем          0
Name: family_status_id, dtype: int64

We see that each specific status corresponds to a specific id. In order to facilitate the dataframe, you can put redundant data in a separate table and remove duplicates from it.

In [116]:
family_status_dict = df[['family_status','family_status_id']]
family_status_dict = family_status_dict.drop_duplicates().reset_index(drop=True)
print(family_status_dict)

           family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4


You can delete the **family_status** column from the main df table:

In [117]:
del df['family_status']
display(df.head())

children  dob_years education  education_id  family_status_id gender  \
0         1       42.0    высшее             0                 0      F   
1         1       36.0   среднее             1                 0      F   
2         0       33.0   Среднее             1                 0      M   
3         3       32.0   среднее             1                 0      M   
4         0       53.0   среднее             1                 1      F   

  income_type  debt   total_income                     purpose  \
0   сотрудник     0  253875.639453               покупка жилья   
1   сотрудник     0  112080.014102     приобретение автомобиля   
2   сотрудник     0  145885.952297               покупка жилья   
3   сотрудник     0  267628.550329  дополнительное образование   
4   пенсионер     0  158616.077870             сыграть свадьбу   

   abs_days_employed  
0        8437.673028  
1        4024.803754  
2        5623.422610  
3        4124.747207  
4      340266.072047

The same should be done with columns about education, provided that there are no discrepancies in the data and a strictly defined id corresponds to a certain value for education. To do this, we will output the average id for each value, since the slightest deviation from the integer will show the presence of different IDs for the same level of education:

In [118]:
df.groupby('education')['education_id'].mean()

education
ВЫСШЕЕ                 0
Высшее                 0
НАЧАЛЬНОЕ              3
НЕОКОНЧЕННОЕ ВЫСШЕЕ    2
Начальное              3
Неоконченное высшее    2
СРЕДНЕЕ                1
Среднее                1
УЧЕНАЯ СТЕПЕНЬ         4
Ученая степень         4
высшее                 0
начальное              3
неоконченное высшее    2
среднее                1
ученая степень         4
Name: education_id, dtype: int64

All numbers are integers, so a specific id corresponds to a specific level of education. At the same time, there are hidden duplicates, which can be disposed of in a separate table:

In [119]:
pd.options.mode.chained_assignment = None
education_dict = df[['education','education_id']]
education_dict['education_lowercase'] = education_dict['education'].str.lower()
education_dict = education_dict.groupby('education_lowercase')['education_id'].mean().sort_values()
display(education_dict)

education_lowercase
высшее                 0
среднее                1
неоконченное высшее    2
начальное              3
ученая степень         4
Name: education_id, dtype: int64

You can delete the **education** column from the main df table:

In [120]:
del df['education']
display(df.head())

children  dob_years  education_id  family_status_id gender income_type  \
0         1       42.0             0                 0      F   сотрудник   
1         1       36.0             1                 0      F   сотрудник   
2         0       33.0             1                 0      M   сотрудник   
3         3       32.0             1                 0      M   сотрудник   
4         0       53.0             1                 1      F   пенсионер   

   debt   total_income                     purpose  abs_days_employed  
0     0  253875.639453               покупка жилья        8437.673028  
1     0  112080.014102     приобретение автомобиля        4024.803754  
2     0  145885.952297               покупка жилья        5623.422610  
3     0  267628.550329  дополнительное образование        4124.747207  
4     0  158616.077870             сыграть свадьбу      340266.072047

The final stage of error handling is to count explicit duplicates:

In [121]:
print("Number of explicit duplicates: {}".format(df.duplicated().sum()))

Number of explicit duplicates: 72


Let's remove the explicit duplicates:

In [122]:
df = df.drop_duplicates().reset_index(drop=True)

Let's check that the explicit duplicates are removed:

In [123]:
#print("Number of explicit duplicates: {}".format(df.duplicated().sum()))
print(f"Number of explicit duplicates: {df.duplicated().sum()}")

Number of explicit duplicates: 0


### Categorization

To answer the question about the dependence of the presence of children and the repayment of the loan on time, it is necessary to add a column in which information about the very fact of having children, and not about their number, will be reflected. To do this, write the appropriate function and apply it:

In [124]:
def has_children(children):
    if children >= 1:
        return 'there are children'
    else:
        return 'no children'

In [125]:
df['has_children'] = df['children'].apply(has_children)
df.head(10)

children  dob_years  education_id  family_status_id gender income_type  \
0         1       42.0             0                 0      F   сотрудник   
1         1       36.0             1                 0      F   сотрудник   
2         0       33.0             1                 0      M   сотрудник   
3         3       32.0             1                 0      M   сотрудник   
4         0       53.0             1                 1      F   пенсионер   
5         0       27.0             0                 1      M   компаньон   
6         0       43.0             0                 0      F   компаньон   
7         0       50.0             1                 0      M   сотрудник   
8         2       35.0             0                 1      F   сотрудник   
9         0       41.0             1                 0      M   сотрудник   

   debt   total_income                     purpose  abs_days_employed  \
0     0  253875.639453               покупка жилья        8437.673028   
1     0  112080.014102     приобретение автомобиля        4024.803754   
2     0  145885.952297               покупка жилья        5623.422610   
3     0  267628.550329  дополнительное образование        4124.747207   
4     0  158616.077870             сыграть свадьбу      340266.072047   
5     0  255763.565419               покупка жилья         926.185831   
6     0  240525.971920           операции с жильем        2879.202052   
7     0  135823.934197                 образование         152.779569   
8     0   95856.832424       на проведение свадьбы        6929.865299   
9     0  144425.938277     покупка жилья для семьи        2188.756445   

  has_children  
0    есть дети  
1    есть дети  
2    детей нет  
3    есть дети  
4    детей нет  
5    детей нет  
6    детей нет  
7    детей нет  
8    есть дети  
9    детей нет

The function worked correctly.

Next, let's study the column **total_income**. The column contains data in the format of real numbers. For more accurate calculations in the future, we will need to translate the data into the format of integers.

In [126]:
df['total_income'] = df['total_income'].astype('int')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   children           21453 non-null  int64  
 1   dob_years          21453 non-null  float64
 2   education_id       21453 non-null  int64  
 3   family_status_id   21453 non-null  int64  
 4   gender             21453 non-null  object 
 5   income_type        21453 non-null  object 
 6   debt               21453 non-null  int64  
 7   total_income       21453 non-null  int64  
 8   purpose            21453 non-null  object 
 9   abs_days_employed  21453 non-null  float64
 10  has_children       21453 non-null  object 
dtypes: float64(2), int64(5), object(4)
memory usage: 1.8+ MB


Видим, что формат успешно изменен. 

For more convenient data analysis, it is better to divide monthly income into categories. To determine the optimal range of categories, let's study the range of values in the column **total_income**:

In [127]:
display(df['total_income'].min())
display(df['total_income'].max())
display(df['total_income'].mean())
display(df['total_income'].median())

20667

2265604

165226.54262807066

145017.0

It can be seen that the minimum value is *20667*, and the maximum is *2265604*, while the average and median are in the range *145017 - 165209* . It can be assumed that the income of most people is in the range of *100000-2001*. We will add ±*50000* from the upper and lower assumed boundaries for a wider coverage and write the function: 

In [128]:
def income_range(total_income):
    if total_income < 50000:
        return '< 50k'
    elif total_income <= 100000:
        return '50-100k'
    elif total_income <= 150000:
        return '100-150k'
    elif total_income <=200000:
        return '150-200k'
    elif total_income <=250000:
        return '200-250k'
    else:
        return '>250k'
check = 159000
print(income_range(check))

150-200k


The function worked correctly. Now let's add a new column of values to which this function will return.

In [129]:
df['total_income_range'] = df['total_income'].apply(income_range)
df.head()

children  dob_years  education_id  family_status_id gender income_type  \
0         1       42.0             0                 0      F   сотрудник   
1         1       36.0             1                 0      F   сотрудник   
2         0       33.0             1                 0      M   сотрудник   
3         3       32.0             1                 0      M   сотрудник   
4         0       53.0             1                 1      F   пенсионер   

   debt  total_income                     purpose  abs_days_employed  \
0     0        253875               покупка жилья        8437.673028   
1     0        112080     приобретение автомобиля        4024.803754   
2     0        145885               покупка жилья        5623.422610   
3     0        267628  дополнительное образование        4124.747207   
4     0        158616             сыграть свадьбу      340266.072047   

  has_children total_income_range  
0    есть дети              >250k  
1    есть дети           100-150k  
2    детей нет           100-150k  
3    есть дети              >250k  
4    детей нет           150-200k

Group the data by column **total_income_range** and highlight the proportion of each range to make sure that the categories are set evenly:

In [130]:
total_income_range_df = df.groupby('total_income_range').agg({'debt':'count'})
total_income_range_df['percentage'] = (total_income_range_df['debt']/df['debt'].count())*100
total_income_range_df

debt  percentage
total_income_range                  
100-150k            7806   36.386519
150-200k            4118   19.195451
200-250k            2254   10.506689
50-100k             4091   19.069594
< 50k                372    1.734023
>250k               2812   13.107724

Let's sort the categories as income increases for clarity:

In [131]:
total_income_range_df['order'] = [3,4,5,2,1,6]
total_income_range_df = total_income_range_df.sort_values(by='order')
del total_income_range_df['order']
total_income_range_df

debt  percentage
total_income_range                  
< 50k                372    1.734023
50-100k             4091   19.069594
100-150k            7806   36.386519
150-200k            4118   19.195451
200-250k            2254   10.506689
>250k               2812   13.107724

It can be seen that the data is distributed into categories without strong distortions in the extreme ranges, so the categories are defined correctly. 

**At the same time, it should be noted that the share of borrowers with an income level of less than 50,000 is less than 2% of the entire sample, which does not allow us to draw conclusions for this group due to insufficient data**.

### Lemmatization

The next step is to analyze the data on the purpose of obtaining a loan in the **purpose** column. Let's analyze the data:

In [132]:
df['purpose'].value_counts()

свадьба                                   791
на проведение свадьбы                     767
сыграть свадьбу                           765
операции с недвижимостью                  675
покупка коммерческой недвижимости         661
операции с жильем                         652
покупка жилья для сдачи                   651
операции с коммерческой недвижимостью     650
покупка жилья                             646
жилье                                     646
покупка жилья для семьи                   638
строительство собственной недвижимости    635
недвижимость                              633
операции со своей недвижимостью           627
строительство жилой недвижимости          624
покупка недвижимости                      621
покупка своего жилья                      620
строительство недвижимости                619
ремонт жилью                              607
покупка жилой недвижимости                606
на покупку своего автомобиля              505
заняться высшим образованием      

From the data obtained, it can be seen that the most frequently mentioned goals of obtaining a loan are: 

* **purchase of real estate**,
* **buying a car**,
* **getting an education**,
* **conducting a wedding**.

It is also seen that there are a lot of hidden duplicates in the column due to the different wording of the purpose of the loan received by borrowers. It is necessary to bring this list to uniformity of formulations, for this we use lemmatization:

In [133]:
from pymystem3 import Mystem
m = Mystem()

lemma_string = str(df['purpose'])
lemma = m.lemmatize(lemma_string)

For a more convenient application of lemmatization, we write the function:

In [134]:
def get_purpose_group(purpose):
    lemma = m.lemmatize(purpose)
    if ('недвижимость' in lemma) or ('жилье' in lemma) or ('строительство' in lemma):
        return 1
    elif 'автомобиль' in lemma:
        return 2
    elif 'образование' in lemma:
        return 3
    elif 'свадьба' in lemma:
        return 4
    else:
        return 5

Let's create a new column in the table, which, using the written function, assign a category and display the table on the screen for verification:

In [135]:
df['purpose_id'] = df['purpose'].apply(get_purpose_group)
df.head()

children  dob_years  education_id  family_status_id gender income_type  \
0         1       42.0             0                 0      F   сотрудник   
1         1       36.0             1                 0      F   сотрудник   
2         0       33.0             1                 0      M   сотрудник   
3         3       32.0             1                 0      M   сотрудник   
4         0       53.0             1                 1      F   пенсионер   

   debt  total_income                     purpose  abs_days_employed  \
0     0        253875               покупка жилья        8437.673028   
1     0        112080     приобретение автомобиля        4024.803754   
2     0        145885               покупка жилья        5623.422610   
3     0        267628  дополнительное образование        4124.747207   
4     0        158616             сыграть свадьбу      340266.072047   

  has_children total_income_range  purpose_id  
0    есть дети              >250k           1  
1    есть дети           100-150k           2  
2    детей нет           100-150k           1  
3    есть дети              >250k           3  
4    детей нет           150-200k           4

Let's create a table with categories for different purposes of obtaining a loan:

In [136]:
columns_title = ['purpose','purpose_id']
purpose_data=[['приобретение недвижимости',1],
              ['покупка автомобиля',2],
              ['получение образования',3],
              ['проведение свадьбы',4],
              ['иное',5]
             ]
purpose_dict = pd.DataFrame(data=purpose_data, columns=columns_title)
purpose_dict

purpose  purpose_id
0  приобретение недвижимости           1
1         покупка автомобиля           2
2      получение образования           3
3         проведение свадьбы           4
4                       иное           5

Теперь можно удалить столбец **purpose** из общей таблицы:

In [137]:
df = df.drop('purpose', axis=1)
df.head()

children  dob_years  education_id  family_status_id gender income_type  \
0         1       42.0             0                 0      F   сотрудник   
1         1       36.0             1                 0      F   сотрудник   
2         0       33.0             1                 0      M   сотрудник   
3         3       32.0             1                 0      M   сотрудник   
4         0       53.0             1                 1      F   пенсионер   

   debt  total_income  abs_days_employed has_children total_income_range  \
0     0        253875        8437.673028    есть дети              >250k   
1     0        112080        4024.803754    есть дети           100-150k   
2     0        145885        5623.422610    детей нет           100-150k   
3     0        267628        4124.747207    есть дети              >250k   
4     0        158616      340266.072047    детей нет           150-200k   

   purpose_id  
0           1  
1           2  
2           1  
3           3  
4           4

In [138]:
#del df['purpose']
#df.head()

With this action, you can finish the data preprocessing stage.

As a result of data preprocessing, a dataframe with a database of borrowers was obtained, including information on a number of key indicators for analysis:

   * having children
   * marital status (in the format of category indexes)
   * income level (in the format of ranges of categories)
   * loan objectives (in the format of category indexes)

Auxiliary tables were also created:
    
   * family_status_dict (with indexes of categories about marital status)
   * purpose_dict (with category indexes about the purpose of the loan)

The prepared data will help answer the questions posed.

___
## Hypothesis testing
___

We will check the availability of dependence on the availability of children and the repayment of the loan on time. To do this, we will build a summary table based on the data we are interested in:

In [140]:
df_pivot_children = df.pivot_table(index='has_children',values='debt',aggfunc='mean')
df_pivot_children['debt'] = pd.Series(["{0:.2f}%".format(val * 100) for val in df_pivot_children['debt']], index = df_pivot_children.index)
df_pivot_children

debt
has_children       
детей нет     7.54%
есть дети     9.21%

According to the results of this comparison, it can be seen that among borrowers with children, the proportion of those who had debt on repayment of loans is higher than among borrowers without children by about 1.67%.
Thus, it can be confirmed that 

**there is a relationship between the presence of children and the repayment of the loan on time**.
* * *

Let's check whether there is a relationship between marital status and repayment of the loan on time. To do this, we will also build a summary table based on the relevant data:

In [45]:
df_pivot_family = df.pivot_table(index='family_status_id',values='debt',aggfunc='mean')
df_pivot_family['debt'] = pd.Series(["{0:.2f}%".format(val * 100) for val in df_pivot_family['debt']], index = df_pivot_family.index)
df_pivot_family

debt
family_status_id       
0                 7.55%
1                 9.35%
2                 6.57%
3                 7.11%
4                 9.75%

Next, we will attach this table to the table with the designation of category indexes to understand the status of each group:

In [46]:
family_status_dict.merge(df_pivot_family,on='family_status_id',how='right')

family_status  family_status_id   debt
0        женат / замужем                 0  7.55%
1       гражданский брак                 1  9.35%
2         вдовец / вдова                 2  6.57%
3              в разводе                 3  7.11%
4  Не женат / не замужем                 4  9.75%

Based on the results of this comparison, it can be seen that the average values for the availability of debt on repayment of loans between groups of borrowers with different marital status varies in the range from 6.57% to 9.75%. The lowest value for the presence of debt is typical for people with the status "widower / widow", borrowers with the status "not married / not married" have the highest indicator. Thus, it can be argued that

**there is a relationship between marital status and repayment of the loan on time**.
* * *

Next, we will check whether there is a relationship between the income level and the repayment of the loan on time. To do this, we will also build a summary table based on the relevant data:

In [47]:
df_pivot_income = df.pivot_table(index='total_income_range',values='debt',aggfunc='mean')
df_pivot_income['debt'] = pd.Series(["{0:.2f}%".format(val * 100) for val in df_pivot_income['debt']], index = df_pivot_income.index)
df_pivot_income

debt
total_income_range       
100-150k            8.47%
150-200k            8.94%
200-250k            7.28%
50-100k             8.09%
< 50k               6.18%
>250k               6.90%

Let's sort the table in ascending order of income level for greater clarity:

In [48]:
df_pivot_income['order'] = [3,4,5,2,1,6]
df_pivot_income = df_pivot_income.sort_values(by='order')
del df_pivot_income['order']
df_pivot_income

debt
total_income_range       
< 50k               6.18%
50-100k             8.09%
100-150k            8.47%
150-200k            8.94%
200-250k            7.28%
>250k               6.90%

According to the results of this comparison, it can be seen that the smallest proportion of borrowers who had debt on repayment of the loan 
it is typical for borrowers with a minimum and maximum income level (up to 50,000 and 200,000 and above). 
At the same time, borrowers with an average income level of 50000-2001 in the sample had the highest values for the presence of
debt. 

At the same time, we remember that the share of borrowers with incomes below 50,000 is less than 2% of the entire sample, which
is not a representative value for the statement about the presence of dependence for this group.

As a result, it can be argued that the lowest indicators for the presence of debt are typical for borrowers with  
high income - from 200,000 and above. Thus, it is true that

**there is a dependence between the income level and the repayment of the loan on time**.
* * *

We will also check how the different purposes of the loan affect its repayment on time. To do this, we will also build a summary table based on the relevant data:

In [49]:
df_pivot_purpose = df.pivot_table(index='purpose_id',values='debt',aggfunc='mean')
df_pivot_purpose['debt'] = pd.Series(["{0:.2f}%".format(val * 100) for val in df_pivot_purpose['debt']], index = df_pivot_purpose.index)
df_pivot_purpose

debt
purpose_id       
1           7.23%
2           9.36%
3           9.22%
4           8.01%

Next, we will attach this table to the table with the designation of category indexes to understand the goals of each group:

In [50]:
purpose_dict.merge(df_pivot_purpose,on='purpose_id',how='right')

purpose  purpose_id   debt
0  приобретение недвижимости           1  7.23%
1         покупка автомобиля           2  9.36%
2      получение образования           3  9.22%
3         проведение свадьбы           4  8.01%

Based on the results of this comparison, it can be seen that the average values for the availability of debt on repayment of loans between groups of borrowers with different purposes of obtaining a loan varies in the range from 7.23% to 9.35%. The lowest value for the presence of debt is typical for people taking a loan for "real estate purchase", borrowers taking a loan for "car purchase" have the highest indicator. Thus, it can be confirmed that

**the purpose of obtaining a loan affects its repayment on time**.
* * *

**Conclusions**

All hypotheses were confirmed:
* there is a relationship between the presence of children and the repayment of the loan on time
* there is a relationship between marital status and repayment of the loan on time
* there is a relationship between the income level and the repayment of the loan on time
* the purpose of obtaining a loan affects its repayment on time
_ _ _ 

## Results of the study

Four hypotheses were tested and it was confirmed:
* Borrowers without children were less likely to admit the appearance of debt on loan repayment.
* Borrowers with the status of marital status "civil marriage" and "not married / not married" most often had  
debt on loan repayment on time..
* The income level of borrowers and the repayment of the loan on time also have a relationship, but this hypothesis was partially confirmed
due to the small number of low-income borrowers. People with a high income level (from 200,000) were less likely to admit
the appearance of debt on loan repayment on time.
* The purpose of obtaining a loan affects its repayment on time: people who took a loan to buy real estate and conduct  
weddings more often gave a loan on time, while for people with the purpose of obtaining a loan "buying a car" и 
"getting an education" is characterized by a higher indicator of the presence of debt when the loan is repaid.